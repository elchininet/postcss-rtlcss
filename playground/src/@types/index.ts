export type NumberOrString = string | number;
export type NumberOrStringArray = NumberOrString[];

export interface ObjectProps<T> {
    [key: string]: T;
}