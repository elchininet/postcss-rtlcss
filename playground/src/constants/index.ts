export enum COLORS {
    action = '#18AEC9',
    active = '#026C7F',
    gray = '#DDD',
    gray_dark = '#555',
    gray_darkest = '#1E1E1E',
    gray_light = '#EFEFEF',
    white = '#FFF'
}