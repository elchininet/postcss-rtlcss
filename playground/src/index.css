html,
body {
    height: 100%;
    margin: 0;
    padding: 0;
}

body {
    font-family: Arial, Helvetica, sans-serif;
    overscroll-behavior-y: contain;
}

#app {
    height: 100%;
}