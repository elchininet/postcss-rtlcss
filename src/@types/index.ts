import { Rule, AtRule } from 'postcss';

export interface ObjectWithProps<T> {
    [key: string]: T;
}

export enum Mode {
    combined = 'combined',
    override = 'override'
}

export enum Source {
    ltr = 'ltr',
    rtl = 'rtl'
}

export enum Autorename {
    disabled = 'disabled',
    flexible = 'flexible',
    strict = 'strict'
}

export type ModeValues = keyof typeof Mode;
export type SourceValues = keyof typeof Source;
export type AutorenameValues = keyof typeof Autorename;

export type strings = string | string[];

interface StringMapOptions {
    scope: '*' | 'selector' | 'url';
    ignoreCase: boolean;
}

export interface StringMap {
    name: string;
    priority: number;
    search: strings;
    replace: strings;
    options: StringMapOptions;
}

export interface PluginStringMap {
    name?: string;
    search: strings;
    replace: strings;
}

export interface PluginOptions {
    mode?: ModeValues;
    ltrPrefix?: strings;
    rtlPrefix?: strings;
    bothPrefix?: strings;
    source?: SourceValues;
    processUrls?: boolean;
    processKeyFrames?: boolean;
    useCalc?: boolean;
    stringMap?: PluginStringMap[];
    autoRename?: Autorename;
    greedy?: boolean;
}

export interface PluginOptionsNormalized extends Omit<Required<PluginOptions>, 'stringMap'> {
    stringMap: StringMap[];
}

export interface RulesObject {
    rule: Rule;
    ruleLTR: Rule | null;
    ruleRTL: Rule | null;
    ruleBoth: Rule | null;
}

export interface AtRulesObject {
    atRuleParams: string;
    atRule: AtRule;
    atRuleFlipped: AtRule;
}

export interface AtRulesStringMapObject {
    name: string;
    nameFlipped: string;
}

export interface AtRulesStringMap {
    [key: string]: AtRulesStringMapObject;
}

export interface ShortHandsData {
    [key: string]: {
        overridden: string | null;
        overrides: string[];
    };
}