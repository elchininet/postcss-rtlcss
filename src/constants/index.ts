export const COMMENT_TYPE = 'comment';
export const DECLARATION_TYPE = 'decl';
export const RULE_TYPE = 'rule';
export const AT_RULE_TYPE = 'atrule';
export const STRING_TYPE = 'string';
export const BOOLEAN_TYPE = 'boolean';
export const KEYFRAMES_NAME = 'keyframes';
export const ANIMATION_PROP = 'animation';
export const ANIMATION_NAME_PROP = 'animation-name';
export const RTL_COMMENT_REGEXP = /rtl:/;
export const RTL_CONTROL_DIRECTIVE_REG_EXP = /^\/\*!? *rtl:?(\w+)?:(\w+):?([^*]*?) *\*\/$/;
export const FLIP_PROPERTY_REGEXP = /(right|left)/i;

export enum CONTROL_DIRECTIVE {
    IGNORE = 'ignore'
}

export enum CONTROL_DIRECTIVE_BLOCK {
    BEGIN = 'begin',
    END = 'end'
}