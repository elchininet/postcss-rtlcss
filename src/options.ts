import { Mode, Source } from '@types';

export { Mode, Source };