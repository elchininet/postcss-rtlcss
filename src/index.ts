import postcss, { Root, Transformer } from 'postcss';
import { PluginOptions, RulesObject } from '@types';
import { parseOptions } from '@utilities/options';
import { parseRules } from '@rules/parser';
import { appendRules } from '@utilities/rules';
export { PluginOptions, Mode, Source, PluginStringMap,  } from '@types';

const transformer = (options: PluginOptions = {}): Transformer  => (
    (css: Root): void => {
        const optionsParsed = parseOptions(options);
        const result: RulesObject[] = parseRules(css, optionsParsed);
        appendRules(result);           
    }
);

export const postcssRTLCSS = postcss.plugin<PluginOptions>('postcss-rtlcss', transformer);