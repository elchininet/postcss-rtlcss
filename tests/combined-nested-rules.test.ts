import postcss from 'postcss';
import postcssRTLCSS from '../src';
import { PluginOptions, Mode, Source } from '../src/@types';
import { readCSSFile } from './test-utils';

const baseOptions: PluginOptions = {mode: Mode.combined};

describe('Combined Tests Nested rules', (): void => {

  let input = '';

  beforeEach(async (): Promise<void> => {
    input = input || await readCSSFile('input-nested.scss');
  });

  it('Combined {source: rtl}', (): void => {
    const options: PluginOptions = { ...baseOptions, source: Source.rtl };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Combined {source: ltr}', (): void => {
    const options: PluginOptions = { ...baseOptions, source: Source.ltr };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

});