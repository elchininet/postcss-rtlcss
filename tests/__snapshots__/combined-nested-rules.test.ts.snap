// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combined Tests Nested rules Combined {source: ltr} 1`] = `
".test1 {
    color: red;
    .test2 {
        padding: 5px;
        width: 100%;
    }
}

[dir=\\"ltr\\"] .test1 {
    left: 10px;
    .test2 {
        text-align: left;
    }
}

[dir=\\"rtl\\"] .test1 {
    right: 10px;
    .test2 {
        text-align: right;
    }
}

.test3 {
    &.test4 {
        color: black;
    }
    &.test5 {
        padding: 5px
    }
}

[dir=\\"ltr\\"] .test3 {
    left: 10px;
    &.test4 {
        margin-left: 20px;
    }
    &.test5 {
        span {
            text-align: left;
        }
    }
}

[dir=\\"rtl\\"] .test3 {
    right: 10px;
    &.test4 {
        margin-right: 20px;
    }
    &.test5 {
        span {
            text-align: right;
        }
    }
}

.test6 {
    display: flex;
    .global & {
        color: white;
        div {
            margin: 0 auto;
        }
    }
    .test7 {
        .global & {
            background-color: #FFF;
            ::after {
                content: \\"\\";
                display: block;
            }
        }
    }
}

[dir=\\"ltr\\"] .test6 {
    .global & {
        padding: 5px 10px 5px 5px;
        border-left-color: #666;
        div {
            padding-left: 20px;
        }
    }
    .test7 {
        .global & {
            ::after {
                text-align: left;
            }
        }
    }
}

[dir=\\"rtl\\"] .test6 {
    .global & {
        padding: 5px 5px 5px 10px;
        border-right-color: #666;
        div {
            padding-right: 20px;
        }
    }
    .test7 {
        .global & {
            ::after {
                text-align: right;
            }
        }
    }
}"
`;

exports[`Combined Tests Nested rules Combined {source: rtl} 1`] = `
".test1 {
    color: red;
    .test2 {
        padding: 5px;
        width: 100%;
    }
}

[dir=\\"rtl\\"] .test1 {
    left: 10px;
    .test2 {
        text-align: left;
    }
}

[dir=\\"ltr\\"] .test1 {
    right: 10px;
    .test2 {
        text-align: right;
    }
}

.test3 {
    &.test4 {
        color: black;
    }
    &.test5 {
        padding: 5px
    }
}

[dir=\\"rtl\\"] .test3 {
    left: 10px;
    &.test4 {
        margin-left: 20px;
    }
    &.test5 {
        span {
            text-align: left;
        }
    }
}

[dir=\\"ltr\\"] .test3 {
    right: 10px;
    &.test4 {
        margin-right: 20px;
    }
    &.test5 {
        span {
            text-align: right;
        }
    }
}

.test6 {
    display: flex;
    .global & {
        color: white;
        div {
            margin: 0 auto;
        }
    }
    .test7 {
        .global & {
            background-color: #FFF;
            ::after {
                content: \\"\\";
                display: block;
            }
        }
    }
}

[dir=\\"rtl\\"] .test6 {
    .global & {
        padding: 5px 10px 5px 5px;
        border-left-color: #666;
        div {
            padding-left: 20px;
        }
    }
    .test7 {
        .global & {
            ::after {
                text-align: left;
            }
        }
    }
}

[dir=\\"ltr\\"] .test6 {
    .global & {
        padding: 5px 5px 5px 10px;
        border-right-color: #666;
        div {
            padding-right: 20px;
        }
    }
    .test7 {
        .global & {
            ::after {
                text-align: right;
            }
        }
    }
}"
`;
